tests/python_package_test/test_engine.py

```diff
@@ -4049,6 +4056,27 @@ def test_average_precision_metric():
     assert res["training"]["average_precision"][-1] == pytest.approx(1)
 
 
+def test_r2_metric():
+    # test against sklearn R2 metric
+    X, y = make_synthetic_regression()
+    params = {"objective": "regression", "metric": "r2", "verbose": -1}
+    res = {}
+    train_data = lgb.Dataset(X, label=y)
+    est = lgb.train(
+        params, train_data, num_boost_round=1, valid_sets=[train_data], callbacks=[lgb.record_evaluation(res)]
+    )
+    r2 = res["training"]["r2"][-1]
+    pred = est.predict(X)
+    sklearn_r2 = r2_score(y, pred)
+    assert r2 == pytest.approx(sklearn_r2)
```

Comment:
Can we also assert that `r2 != 0` and `r2 != 1` (edge cases in metric implementation)?